src/kernels/MIOpenCartesianProd.cpp

```diff
@@ -0,0 +1,132 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+template <typename T>
+__device__ void cartesianProdForward(const T* input,
+                                     T* output_ws,
+                                     tensor_view_t<1> input_tv,
+                                     tensor_view_t<2> output_tv,
+                                     size_t stride,
+                                     size_t dim1_idx)
+{
+    size_t gid = threadIdx.x + blockIdx.x * blockDim.x;
+    if(gid >= output_tv.size[0])
+        return;
+    output_ws[output_tv.size[0] * dim1_idx + gid] = input[(gid / stride) % input_tv.size[0]];
+}
+
+extern "C" __global__ void CartesianProdForward(const D_TYPE* __restrict__ input,
+                                                D_TYPE* __restrict__ output_ws,
+                                                tensor_view_t<1> input_tv,
+                                                tensor_view_t<2> output_tv,
+                                                size_t stride,
+                                                size_t dim1_idx)
+{
+    cartesianProdForward<D_TYPE>(input, output_ws, input_tv, output_tv, stride, dim1_idx);
+}
+
+template <typename T>
+__device__ void cartesianProdTranspose(const T* output_ws, T* output, tensor_view_t<2> output_tv)
+{
+    __shared__ T block[TILE_SIZE][TILE_SIZE + 1];
+
+    size_t lid[2];
+    lid[0] = threadIdx.x;
+    lid[1] = threadIdx.y;
+
+    size_t transposed_n[2];
+    transposed_n[1] = blockIdx.x * TILE_SIZE + lid[0];
+    transposed_n[0] = blockIdx.y * TILE_SIZE + lid[1];
+
+    if((transposed_n[1] < output_tv.size[0]) && (transposed_n[0] < output_tv.size[1]))
+    {
+        size_t workspace_idx  = transposed_n[0] * output_tv.size[0] + transposed_n[1];
+        block[lid[1]][lid[0]] = output_ws[workspace_idx];
+    }
+
+    __syncthreads();
+
+    size_t output_n[2];
+    output_n[1] = blockIdx.y * TILE_SIZE + lid[0];
+    output_n[0] = blockIdx.x * TILE_SIZE + lid[1];
+
+    if((output_n[1] < output_tv.size[1]) && (output_n[0] < output_tv.size[0]))
+    {
+        output[output_n[0] * output_tv.size[1] + output_n[1]] = block[lid[0]][lid[1]];
```

Comment:
This looking wrong, in moDNN:
```c++
  output[TV2D_IDX(output_tv, output_n[0], output_n[1])] = block[lid[0]][lid[1]];
```
```suggestion
        output[output_tv.get_tensor_view_idx({output_n[0], output_n[1]})] = block[lid[0]][lid[1]];
```

Reply:
resolved